File: SexyBiscuit.Engine/Chibi/ChibiAnimator.cs
using Microsoft.Xna.Framework;
using SexyBiscuit.Engine.Core;

namespace SexyBiscuit.Engine.Chibi;

/// <summary>
/// Plays procedural and keyed chibi clips, cross-fading between them.
/// </summary>
/// <remarks>
/// <para>
/// The joints are actors, so animation is nothing more exotic than writing
/// <see cref="Transform3D.LocalEulerAngles"/> on sixteen transforms a frame. Those
/// transforms are looked up once and kept: <see cref="Actor.Transform3D"/> is a component
/// search on every access, which is cheap once and is not cheap sixteen times a frame per
/// character.
/// </para>
/// <para>
/// This deliberately has nothing to do with <see cref="Animation.SkeletalAnimator"/>. That
/// drives a bone palette for a skinning shader, has no browser counterpart, and cannot be
/// saved in a scene file — going near it would cost MakeChibi its second engine.
/// </para>
/// <para>Mirrored in <c>html5/src/chibi/ChibiAnimator.js</c>.</para>
/// </remarks>
public sealed class ChibiAnimator : Component
{
    /// <summary>One clip in flight: which clip, how far through, and whether it is done.</summary>
    private sealed class Playhead
    {
        public Playhead(ChibiClip clip) => Clip = clip;

        public ChibiClip Clip { get; }
        public float Time { get; private set; }
        public bool Finished { get; private set; }

        public void Advance(float dt, float speed)
        {
            Time += dt * speed;
            float duration = MathF.Max(1e-4f, Clip.Duration);
            if (Time < duration) return;

            if (Clip.Loop) Time %= duration;
            else { Time = duration; Finished = true; }
        }

        /// <summary>Normalised position through the clip, 0 to 1.</summary>
        public float Phase => MathF.Min(1f, Time / MathF.Max(1e-4f, Clip.Duration));
    }

    /// <summary>The clip to play. An unknown name leaves the character at rest.</summary>
    public string Clip { get; set; } = "idle";

    /// <summary>Playback rate.</summary>
    public float Speed { get; set; } = 1f;

    /// <summary>Scales a procedural clip's swing; keyed clips ignore it.</summary>
    public float Intensity { get; set; } = 1f;

    /// <summary>Whether a clip is running.</summary>
    public bool Playing { get; set; } = true;

    /// <summary>Seconds to cross-fade when the clip changes.</summary>
    public float BlendTime { get; set; } = 0.18f;

    /// <summary>Fires with the clip's name when a non-looping clip reaches its end.</summary>
    public event Action<string>? ClipFinished;

    private readonly Dictionary<string, Transform3D> _joints = new(StringComparer.Ordinal);
    private Transform3D? _root;
    private Vector3 _rootRest;

    private Playhead? _current;
    private Playhead? _previous;
    private float _blend;
    private float _blendLeft;

    private readonly ChibiPose _poseA = new();
    private readonly ChibiPose _poseB = new();
    private readonly ChibiPose _poseOut = new();
    private bool _resolved;

    /// <summary>Every clip name this animator will accept.</summary>
    public static IReadOnlyList<string> Clips => ChibiClips.Names;

    /// <inheritdoc />
    public override void Start()
    {
        Resolve();
        if (Playing) Play(Clip, 0f);
    }

    /// <summary>
    /// Finds the joints to drive. Called automatically, and again by anything that rebuilds
    /// the body underneath — the joints it cached are destroyed actors.
    /// </summary>
    public bool Resolve()
    {
        _joints.Clear();
        _root = null;

        var character = GetComponent<ChibiCharacter>();
        if (character?.Chibi != null)
        {
            foreach (var (name, transform) in character.Chibi.Joints) _joints[name] = transform;
            _root = character.Chibi.Actor.GetComponent<Transform3D>();
        }
        else if (Actor != null)
        {
            // No component to ask: find the joints by name in whatever subtree is there,
            // so a baked-out chibi still animates.
            foreach (Actor actor in Actor.Descendants())
            {
                if (!ChibiParts.JointNames.Contains(actor.Name) || _joints.ContainsKey(actor.Name)) continue;
                if (actor.GetComponent<Transform3D>() is { } transform) _joints[actor.Name] = transform;
            }
        }

        if (_root != null) _rootRest = _root.LocalPosition;
        _resolved = _joints.Count > 0;
        return _resolved;
    }

    /// <summary>Starts a clip, cross-fading from whatever was playing.</summary>
    /// <param name="name">A procedural or keyed clip name.</param>
    /// <param name="blend">Seconds to fade over, or null for <see cref="BlendTime"/>.</param>
    public bool Play(string name, float? blend = null)
    {
        ChibiClip? clip = ChibiClips.Find(name);
        if (clip == null) return false;
        if (_current?.Clip.Name == name && Playing) return true;

        _previous = _current;
        _current = new Playhead(clip);
        _blend = _previous != null ? MathF.Max(0f, blend ?? BlendTime) : 0f;
        _blendLeft = _blend;
        Clip = name;
        Playing = true;
        return true;
    }

    /// <summary>Stops, returning every joint to rest.</summary>
    public void Stop()
    {
        Playing = false;
        _current = null;
        _previous = null;
        ApplyPose(_poseOut.Clear());
    }

    /// <inheritdoc />
    public override void Update(float dt)
    {
        if (!_resolved && !Resolve()) return;
        if (!Playing || _current == null) return;

        _current.Advance(dt, Speed);

        _poseB.Clear();
        _current.Clip.Sample(_current.Phase, _poseB, Intensity);
        ChibiPose pose = _poseB;

        if (_previous != null && _blend > 0f)
        {
            _blendLeft = MathF.Max(0f, _blendLeft - dt);
            float amount = 1f - _blendLeft / _blend;

            _poseA.Clear();
            _previous.Clip.Sample(_previous.Phase, _poseA, Intensity);
            pose = ChibiPose.Blend(_poseOut, _poseA, _poseB, amount);

            if (_blendLeft <= 0f) _previous = null;
        }

        ApplyPose(pose);

        if (!_current.Finished) return;

        ChibiClip finished = _current.Clip;
        if (!finished.Hold) _current = null;
        Playing = finished.Hold;
        ClipFinished?.Invoke(finished.Name);
    }

    /// <summary>
    /// Writes a pose onto the joints.
    /// </summary>
    /// <remarks>
    /// Joints the clip did not mention go back to rest rather than keeping whatever the
    /// last clip left there — a wave must not leave the legs mid-stride.
    /// </remarks>
    public void ApplyPose(ChibiPose pose)
    {
        foreach (string name in ChibiParts.JointNames)
        {
            if (_joints.TryGetValue(name, out Transform3D? transform))
                transform.LocalEulerAngles = pose.Get(name);
        }

        if (_root != null) _root.LocalPosition = _rootRest + pose.Offset;
    }
}

File: html5/src/chibi/ChibiAnimator.js
// -----------------------------------------------------------------------------
// ChibiAnimator — plays a clip over a chibi's joints.
//
// The joints are actors, so animation is nothing more exotic than writing
// `localEulerAngles` on sixteen transforms a frame. Those transforms are looked
// up once and kept: `Actor.transform3D` is a component search on every access,
// which is cheap once and is not cheap sixteen times a frame per character.
//
// This deliberately has nothing to do with SkeletalAnimator. That drives a bone
// palette for a skinning shader, has no browser counterpart, and cannot be saved
// in a scene file — going near it would cost MakeChibi its second engine.
//
// Mirrored in SexyBiscuit.Engine/Chibi/ChibiAnimator.cs.
// -----------------------------------------------------------------------------

import { Component } from '../core/Component.js';
import { registerComponent } from '../core/TypeRegistry.js';
import { PropertyType as P } from '../core/PropertyTypes.js';
import { Transform3D } from '../core/Transform3D.js';
import { Vector3 } from '../math/index.js';
import { JOINT_NAMES } from './ChibiRig.js';
import { Pose, blendInto, findClip, clipNames } from './ChibiClips.js';
import { ChibiCharacter } from './ChibiCharacter.js';

/** One clip in flight: which clip, how far through, and how fast. */
class Playhead {
    constructor(clip) {
        this.clip = clip;
        this.time = 0;
        this.finished = false;
    }

    advance(dt, speed) {
        this.time += dt * speed;
        const duration = Math.max(1e-4, this.clip.duration);
        if (this.time < duration) return;

        if (this.clip.loop) this.time %= duration;
        else { this.time = duration; this.finished = true; }
    }

    /** Normalised position through the clip, 0 to 1. */
    get phase() { return Math.min(1, this.time / Math.max(1e-4, this.clip.duration)); }
}

/** Plays procedural and keyed chibi clips, cross-fading between them. */
export class ChibiAnimator extends Component {
    static schema = {
        clip: { type: P.String, default: 'idle' },
        speed: { type: P.Number, default: 1 },
        intensity: { type: P.Number, default: 1 },
        playing: { type: P.Bool, default: true },
        blendTime: { type: P.Number, default: 0.18 },
    };

    constructor() {
        super();

        /** The clip to play. Unknown names leave the character at rest. */
        this.clip = 'idle';
        /** Playback rate. */
        this.speed = 1;
        /** Scales a procedural clip's swing; keyed clips ignore it. */
        this.intensity = 1;
        this.playing = true;
        /** Seconds to cross-fade when the clip changes. */
        this.blendTime = 0.18;

        /** Fires with the clip's name when a non-looping clip reaches its end. */
        this.onClipFinished = null;

        /** @type {Map<string, Transform3D>} Resolved once, in start. */
        this._joints = new Map();
        /** @type {?Transform3D} The chibi root, which carries a clip's offset. */
        this._root = null;
        this._rootRest = new Vector3(0, 0, 0);

        this._current = null;
        this._previous = null;
        this._blend = 0;

        this._poseA = new Pose();
        this._poseB = new Pose();
        this._poseOut = new Pose();
        this._resolved = false;
    }

    /** Every clip name this animator will accept. */
    static get clips() { return clipNames(); }

    start() {
        this.resolve();
        if (this.playing) this.play(this.clip, 0);
    }

    /**
     * Finds the joints to drive. Called automatically, and again by anything that
     * rebuilds the body underneath — the joints it cached are destroyed actors.
     */
    resolve() {
        this._joints.clear();
        this._root = null;

        const character = this.getComponent(ChibiCharacter);
        if (character?.chibi) {
            for (const [name, transform] of character.chibi.joints) this._joints.set(name, transform);
            this._root = character.chibi.actor.getComponent(Transform3D);
        } else {
            // No component to ask: find the joints by name in whatever subtree is
            // there, so a baked-out chibi still animates.
            for (const actor of this.actor.descendants()) {
                if (JOINT_NAMES.includes(actor.name) && !this._joints.has(actor.name)) {
                    this._joints.set(actor.name, actor.getComponent(Transform3D));
                }
            }
        }

        if (this._root) this._rootRest = this._root.localPosition.clone();
        this._resolved = this._joints.size > 0;
        return this._resolved;
    }

    /**
     * Starts a clip, cross-fading from whatever was playing.
     *
     * @param {string} name A procedural or keyed clip name.
     * @param {number} [blend] Seconds to fade over; defaults to `blendTime`.
     */
    play(name, blend = null) {
        const clip = findClip(name);
        if (!clip) return false;
        if (this._current?.clip.name === name && this.playing) return true;

        this._previous = this._current;
        this._current = new Playhead(clip);
        this._blend = this._previous ? Math.max(0, blend ?? this.blendTime) : 0;
        this._blendLeft = this._blend;
        this.clip = name;
        this.playing = true;
        return true;
    }

    /** Stops, returning every joint to rest. */
    stop() {
        this.playing = false;
        this._current = null;
        this._previous = null;
        this.applyPose(this._poseOut.clear());
    }

    update(dt) {
        if (!this._resolved && !this.resolve()) return;
        if (!this.playing || !this._current) return;

        this._current.advance(dt, this.speed);

        this._poseB.clear();
        this._current.clip.sample(this._current.phase, this._poseB, this.intensity);

        let pose = this._poseB;

        if (this._previous && this._blend > 0) {
            this._blendLeft = Math.max(0, this._blendLeft - dt);
            const amount = 1 - this._blendLeft / this._blend;

            this._poseA.clear();
            this._previous.clip.sample(this._previous.phase, this._poseA, this.intensity);
            pose = blendInto(this._poseOut, this._poseA, this._poseB, amount);

            if (this._blendLeft <= 0) this._previous = null;
        }

        this.applyPose(pose);

        if (!this._current.finished) return;

        const finished = this._current.clip;
        if (!finished.hold) this._current = null;
        this.playing = finished.hold;
        if (this.onClipFinished) this.onClipFinished(finished.name);
    }

    /**
     * Writes a pose onto the joints. Joints the clip did not mention go back to
     * rest rather than keeping whatever the last clip left there — a wave must
     * not leave the legs mid-stride.
     */
    applyPose(pose) {
        for (const name of JOINT_NAMES) {
            const transform = this._joints.get(name);
            if (!transform) continue;
            const [x, y, z] = pose.get(name);
            transform.localEulerAngles = new Vector3(x, y, z);
        }

        if (!this._root) return;
        this._root.localPosition = new Vector3(
            this._rootRest.x + pose.offset[0],
            this._rootRest.y + pose.offset[1],
            this._rootRest.z + pose.offset[2]);
    }
}
registerComponent(ChibiAnimator, {
    category: 'Chibi',
    summary: 'Plays procedural and keyed clips over a chibi rig.',
});
